File: androidApp/src/main/java/de/hergt/kmm/notes/android/notes/NotesViewModel.kt
package de.hergt.kmm.notes.android.notes

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import de.hergt.kmm.core.TimeUtils
import de.hergt.kmm.core.extensions.isNotNull
import de.hergt.kmm.notes.Notes
import de.hergt.kmm.notes.database.DatabaseFactory
import de.hergt.kmm.notes.database.NotesRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.UUID

class NotesViewModel(context: Context): ViewModel() {

    private val mDatabase = DatabaseFactory(context)
    private val mNotesRepository = NotesRepository(mDatabase.mNotesQueries)

    val mTimeUtil = TimeUtils()

    private val _note: MutableStateFlow<Notes?> = MutableStateFlow(null)
    val selectedNote = _note.asStateFlow()

    fun getAllNotes(): Flow<List<Notes>> {
        return mNotesRepository.getNotes()
    }

    fun upsertNote(note: Notes) {
        viewModelScope.launch {
            withContext(viewModelScope.coroutineContext) { getAllNotes().firstOrNull() }?.let { noteList ->
                if (noteList.firstOrNull { it.mId == note.mId }.isNotNull()) {
                    mNotesRepository.updateNote(note)
                } else {
                    mNotesRepository.createNote(note)
                }
            }
        }
    }

    fun deleteNoteById(id: Long) {
        mNotesRepository.deleteNoteById(id)
    }

    fun selectNewNote() {
        _note.value = Notes(
            UUID.randomUUID().mostSignificantBits,
            null,
            null,
            mTimeUtil.currentTimeMillis()
        )
    }

    fun selectNote(id: Long) {
        viewModelScope.launch {
            getNoteById(id).firstOrNull()?.let {
                _note.value = it
            }
        }
    }

    fun unselectNote() {
        _note.value = null
    }

    fun updateTitle(value: String) {
        _note.value = _note.value?.copy(mTitle = value)
    }

    fun updateText(value: String) {
        _note.value = _note.value?.copy(mText = value)
    }

    private fun getNoteById(id: Long): Flow<Notes> {
        return mNotesRepository.getNoteById(id)
    }

}

File: iosApp/iosApp/notes/NotesViewModel.swift
//
//  NotesViewModel.swift
//  iosApp
//
//  Created by Leon Hergt on 15.04.23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class NotesViewModel: ObservableObject {
    
    @Published var mSelectedNote: Notes? = nil
    @Published var mNotes: [Notes] = []
    
    
    private let mTimeUtils = TimeUtils()
    private let mNotesRepo = NotesRepository(mQueries: DatabaseFactory().mNotesQueries)
    
    
    func getAllNotes() {
        
        upsertNote(note: Notes(mId: UUID().integers.0, mTitle: "Test", mText: "Text...", mDate: mTimeUtils.currentTimeMillis()))
        
        mNotesRepo.getNotes().watch { notes in
            guard let notes = notes as? Array<Notes> else { return }
            self.mNotes = notes
        }
    }
    
    func upsertNote(note: Notes) {
        if mNotes.first(where: { $0.mId == note.mId }) != nil {
            mNotesRepo.updateNote(note: note)
        } else {
            mNotesRepo.createNote(note: note)
        }
    }
    
    func deleteNoteById(id: Int64) {
        mNotesRepo.deleteNoteById(id: id)
    }
    
    func selectNewNote() {
        mSelectedNote = Notes(
            mId: UUID().integers.0,
            mTitle: nil,
            mText: nil,
            mDate: mTimeUtils.currentTimeMillis()
        )
    }
    
    func selectNote(id: Int64) {
        getNoteById(id: id).watch { note in
            guard let note = note else { return }
            self.mSelectedNote = note
        }
    }
    
    func unselectNote() {
        mSelectedNote = nil
    }
    
    func updateTitle(value: String) {
        guard let selectedNote = mSelectedNote else { return }
        mSelectedNote = selectedNote.doCopy(
            mId: selectedNote.mId,
            mTitle: value,
            mText: selectedNote.mText,
            mDate: selectedNote.mDate
        )
    }
    
    func updateText(value: String) {
        guard let selectedNote = mSelectedNote else { return }
        mSelectedNote = selectedNote.doCopy(
            mId: selectedNote.mId,
            mTitle: selectedNote.mTitle,
            mText: value,
            mDate: selectedNote.mDate
        )
    }
    
    private func getNoteById(id: Int64) -> CommonFlow<Notes> {
        return mNotesRepo.getNoteById(id: id)
    }
    
}
